src/main/java/org/folio/service/impl/PatronPinService.java

```diff
@@ -0,0 +1,34 @@
+package org.folio.service.impl;
+
+import static javax.xml.bind.DatatypeConverter.printHexBinary;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+
+import javax.crypto.SecretKey;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PatronPinService {
+  private static final Logger logger = LogManager.getLogger(PatronPinService.class);
+
+  private final PasswordHashService passwordHashService;
+
+  public PatronPinService(PasswordHashService passwordHashService) {
+    this.passwordHashService = passwordHashService;
+  }
+
+  public String derivePin(String pin, String userId) {
```

Comment:
You should probably rename `String userId` to `String salt`.

Reply:
I chose to refer to this as the userId in order to inform the client of the code that it is the `user ID` that is expected, if it is named `salt`, the client has to guess what the expected value is.

I consider the use of the user ID as the salt a decision made within this class. The client does not need to know that is how the user ID is used.

Reply:
ok